### PR TITLE
Sockets refactor: allow any family/type/protocol association

### LINKED ISSUE #3214: UNIXSocket does not work when socket type is DGRAM
Hi !
I'm trying to make a wpa_supplicant front end in pure-crystal, and I need to connect to the server using Unix Socket with the DGRAM (datagram) socket type.

The UNIXSocket class provides a good abstraction, but there are some specificity when using DGRAM that are not handled : 

``` crystal
client = UNIXSocket.new "/path/to/server", Socket::Type::DGRAM
client << "PING"
puts "Result = #{client.gets(4096)}"
client.close
```

This code will hang forever on `client.gets(4096)`. It's not the server fault, I tested with wpa_supplicant server..
[Found the answer on stackoverflow](http://stackoverflow.com/questions/3324619/unix-domain-socket-using-datagram-communication-between-one-server-process-and)

The reason is that the socket is not properly configured when the DGRAM socket type is used : 

> With datagram sockets, specifically unix domain datagram sockets, the client has to bind() to its own endpoint, then connect() to the server's endpoint.

Another exemple is in wpa_cli source (a wpa_supplicant front end) at [this line 93](http://w1.fi/hostapd/devel/wpa__ctrl_8c_source.html#l00093), where it creates the socket that will connect to the server : 
- `socket` : Create the socket
- `bind` : Give the socket a name (<== this is the missing part)
- `connect` : Connect to the server

I've made & push a **_dirty_** [fix here](https://github.com/Bew78LesellB/crystal-wpa_supplicant/blob/master/src/wpa_cri.cr)

I don't really know if my fix is complete and how to implement this in crystal stdlib (naming/logic/..)
@asterite @ysbaddaden : You worked on this, maybe you can help ?

### PR DESCRIPTION
**Pull request by ysbaddaden**
The whole idea behing this refactor is to give more possibilities over sockets, for some less common scenarios (eg: DGRAM UNIX sockets) as well as cleanup the sockets implementation, by avoiding some repetition.

Refactor:

- Socket is now enough to create, configure and use any kind of socket association of family, type and protocol is also possible, as long as it's supported by the underlying OS implementation;
  - as a consequence the `connect`, `bind`, `listen`, `accept`, `send` and `receive` methods were moved from subclasses into `Socket` itself;
- The `TCPSocket`, `TCPServer`, `UDPSocket`, `UNIXSocket` and `UNIXServer` classes are merely sugar to avoid having to deal with socket details;
- `UNIXSocket` and `UNIXServer` can now be used in DGRAM type, in addition to the default STREAM type.

Features:

- `Addrinfo` DNS resolver, that wraps results from `getaddrinfo`;
- `Socket#receive : {String, IPAddress}` to receive a String (like we can `send(String)`;
- `Socket#bind(Int)` to bind to all available interfaces;
- `Socket::Server` module, included by both `TCPServer` and `UNIXServer`;
- `SO_REUSEPORT` socket option; automatically set for `TCPServer` —see http://stackoverflow.com/a/14388707/199791

Breaking Change:

- `IPAddress` now automatically detects the address family, so the argument was removed (limited impact).

Examples:

```crystal
sock = UNIXSocket.new(Socket::Family::DGRAM)
sock.connect("/tmp/service.sock")
sock.send(message)
```

```crystal
sock = Socket.tcp(Socket::Family::INET6)
# set some socket options (e.g. SO_REUSEPORT)
sock.bind(80)
sock.listen
```

closes #3214
closes #3495 
refs #3586

### CODE REVIEW

**Comment by maxpowa** [comment]
Also closes #3495

**Comment by ysbaddaden** [comment]
Thanks, I couldn't find this issue anymore.

File: src/socket.cr [hunk 1]
```diff
@@ -31,110 +31,310 @@ class Socket < IO::FileDescriptor
...
-    def initialize(@family : Family, @address : String, port : Int)
-      if family != Family::INET && family != Family::INET6
-        raise ArgumentError.new("Unsupported address family")
-      end
+  # :nodoc:
+  SOMAXCONN = 128
+
+  property family : Family
+  property type : Type
+  property protocol : Protocol
```
Comment (asterite) [inline src/socket.cr]: Can these three be changed after creating a socket? If not, they should probably be getters.

Comment (asterite) [inline src/socket.cr]: Can these three be changed after creating a socket? If not, they should probably be getters.

File: src/socket.cr [hunk 2]
```diff
@@ -31,110 +31,310 @@ class Socket < IO::FileDescriptor
...
+  protected def initialize(fd : Int32, @family, @type, @protocol = Protocol::IP)
+    init_close_on_exec(fd)
+    super fd, blocking: false
+    self.sync = true
+  end
+
+  # Force opened sockets to be closed on `exec(2)`. Only for platforms that don't
+  # support `SOCK_CLOEXEC` (e.g., Darwin).
+  protected def init_close_on_exec(fd : Int32)
+    {% unless LibC.constants.includes?("SOCK_CLOEXEC".id) %}
```
Comment (asterite) [inline src/socket.cr]: You can also use `LibC.has_constant?("SOCK_CLOEXEC")` (shorter plus probably a bit faster.) Not sure it works on the old compiler without `.id`, so it's probably better to leave the `.id` for now

File: src/socket.cr [hunk 3]
```diff
@@ -31,110 +31,310 @@ class Socket < IO::FileDescriptor
...
-      case family
-      when Family::INET  then LibC::SocklenT.new(sizeof(LibC::SockaddrIn))
-      when Family::INET6 then LibC::SocklenT.new(sizeof(LibC::SockaddrIn6))
-      else                    LibC::SocklenT.new(0)
-      end
+  # Binds the socket on *port* to all local interfaces.
+  #
+  # ```
+  # sock = Socket.tcp(Socket::Family::INET6)
+  # sock.bind "localhost", 1234
```
Comment (asterite) [inline src/socket.cr]: Should this doc comment just be `sock.bind 1234`?

File: src/socket/ip_socket.cr [hunk 4]
```diff
@@ -1,139 +1,25 @@
 class IPSocket < Socket
   # Returns the `IPAddress` for the local end of the IP socket.
   def local_address
-    sockaddr = uninitialized LibC::SockaddrIn6
+    sockaddr = Pointer(LibC::SockaddrIn6).malloc.as(LibC::Sockaddr*)
```
Comment (asterite) [inline src/socket/ip_socket.cr]: Why this was changed? It's an extra allocation (tiny, but still...)

File: src/socket/ip_socket.cr [hunk 5]
```diff
@@ -1,139 +1,25 @@
...
 
-    IPAddress.new(sockaddr, addrlen)
+    IPAddress.from(sockaddr, addrlen)
   end
 
   # Returns the `IPAddress` for the remote end of the IP socket.
   def remote_address
-    sockaddr = uninitialized LibC::SockaddrIn6
-    addrlen = LibC::SocklenT.new(sizeof(LibC::SockaddrIn6))
+    sockaddr = Pointer(LibC::SockaddrIn6).malloc.as(LibC::Sockaddr*)
```
Comment (asterite) [inline src/socket/ip_socket.cr]: Same here, why the change?

**Comment by asterite** [comment]
Looks really good to me! It enables some new stuff, like #3495 and the code looks much nicer, readable and organized. I just made a few comments.

File: src/socket/ip_socket.cr (hunk 4 above)
Comment (ysbaddaden) [inline src/socket/ip_socket.cr]: I ran into some issues. I hope I can change it back. Same below.

File: src/socket.cr (hunk 1 above)
Comment (ysbaddaden) [inline src/socket.cr]: You're right. I'll change that.

File: src/socket/udp_socket.cr [hunk 6]
```diff
@@ -48,113 +51,35 @@ require "./ip_socket"
...
-        next false if addrinfo.ai_next
-        raise err
-      end
-
-      true
-    end
+  def receive(max_message_size = 512) : {String, IPAddress}
+    bytes = Bytes.new(max_message_size)
+    bytes_read, sockaddr, addrlen = recvfrom(bytes)
+    {String.new(bytes.to_unsafe, bytes_read), IPAddress.from(sockaddr, addrlen)}
```
Comment (asterite) [inline src/socket/udp_socket.cr]: Here there's a double allocation: `Bytes.new` and `String.new` (it copies the bytes). You can do something like:

```cr
str = String.new(max_message_size) do |buffer|
  bytes = Slice.new(buffer, max_message_size)
  bytes_read, sockaddr, addrlen = recvfrom(bytes)
  # etc.
  {bytes_read, 0}
end
```

The only "problem" with this is that this will maybe allocate more bytes than needed, so maybe occupy more memory until the string is GC'ed. However, I'll soon change `String#new` to `realloc` the pointer if the resulting bytesize is less than the requested capacity. I even did a few benchmarks and this makes a program go much faster as later less memory needs to be reclaimed.

File: src/socket/ip_socket.cr (hunk 4 above)
Comment (asterite) [inline src/socket/ip_socket.cr]: No worries, we can take care of this later. It's just a tiny allocation and only used if you do request the local address :-)

**Comment by ysbaddaden** [comment]
@asterite I applied all your suggestions, and skipped a few more extra allocations!

I also added support for `SO_REUSEPORT` which is a nice feature, and applied it automatically to `TCPServer` in addition to `SO_REUSEADDR`.

**Comment by asterite** [comment]
@ysbaddaden Oops, my bad! It seems `has_constant?` can be passed a string literal, and it works fine in both the current and next versions of the compiler. So changing those `has_constant?("SOME".id)` to `has_constant?("SOME")` should make travis happy, and then I'll merge this :-)

**Comment by ysbaddaden** [comment]
Fixed! Let's wait for Travis :-)

**Comment by asterite** [comment]
Almost! Just missing `crystal tool format`

**Comment by asterite** [comment]
Ah, never mind. I'll make a release today so i'll merge this and then run the formatter.

@ysbaddaden Thank you so much for this! ❤️
